src/util/triplea/mapXmlCreator/CanalDefinitionsPanel.java

```diff
@@ -0,0 +1,450 @@
+package util.triplea.mapXmlCreator;
+
+import java.awt.Color;
+import java.awt.FontMetrics;
+import java.awt.Graphics;
+import java.awt.Graphics2D;
+import java.awt.Point;
+import java.awt.event.MouseEvent;
+import java.awt.geom.Rectangle2D;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Optional;
+import java.util.Set;
+import java.util.SortedSet;
+import java.util.TreeSet;
+
+import javax.swing.JOptionPane;
+import javax.swing.SwingUtilities;
+
+import com.google.common.collect.Maps;
+import com.google.common.collect.Sets;
+
+import games.strategy.common.swing.SwingAction;
+import util.triplea.mapXmlCreator.TerritoryDefinitionDialog.DEFINITION;
+
+
+final public class CanalDefinitionsPanel extends ImageScrollPanePanel {
+  static final String HTML_CANAL_KEY_POSTFIX = ": ";
+  static final String HTML_CANAL_KEY_PREFIX = "<br/> - ";
+  private static final double PI_HALF = Math.PI / 2;
+  private static final String NEW_CANAL_OPTION = "<new Canal>";
+
+  private Set<String> selectedLandTerritories = new TreeSet<String>();
+  private Set<String> selectedWaterTerritories = new TreeSet<String>();
+  private Optional<String> currentCanalName = Optional.empty();
+
+  private CanalDefinitionsPanel() {}
+
+  public static void layout(final MapXmlCreator mapXMLCreator) {
+    ImageScrollPanePanel.mapXMLCreator = mapXMLCreator;
+    final CanalDefinitionsPanel panel = new CanalDefinitionsPanel();
+    panel.layout(mapXMLCreator.getStepActionPanel());
+    mapXMLCreator.setAutoFillAction(SwingAction.of(e -> {
+      final int prevCanalCount = MapXmlHelper.getCanalDefinitionsMap().size();
+      if (prevCanalCount > 0) {
+        if (JOptionPane.YES_OPTION != MapXmlUIHelper.showYesNoOptionDialog("Auto-Fill Warning",
+            "All current canal definitions will be deleted.\rDo you want to continue with Auto-Fill?",
+            JOptionPane.WARNING_MESSAGE)) {
+          return;
+        }
+        MapXmlHelper.clearCanalDefinitions();
+      }
+      panel.clearSelection();
+
+      final Map<String, Set<String>> landWaterTerritoyConnections = MapXmlData.getLandWaterTerritoryConnections();
+
+      MapXmlHelper.validateAndAddCanalDefinitions(landWaterTerritoyConnections);
+
+      final boolean noNewCanalsBuild = MapXmlHelper.getCanalDefinitionsMap().isEmpty();
+      if (noNewCanalsBuild) {
+        JOptionPane.showMessageDialog(null, "No canals have been build!", "Auto-Fill Result",
```

Comment:
Some copy/paste here, Instead of showing the whole message dialog, can assign to a string variable and then have one line that displays the message, using the approprivate value of the string variable.

Sofor example:
```
final String msg;
if( noNewCanalsBuild) { msg = ...; } else { msg =...;}
JOptoinPane.showMessage...
```